### DISCUSSION TITLE
Buffer size for Bsend

### DISCUSSION THREAD

**Discussion by DirkKuhn** [mpi4py/mpi4py#361] [Q&A] [answered]
Hallo,

what is the correct way to call `Pack_size` when sending multiple messages with `Bsend` in sequence. Maybe the following example illustrates better what I mean. `N` numbers are send from `rank=0` to `rank=1` in a buffered fashion.
```python
import numpy as np
from mpi4py import MPI
from mpi4py.util import dtlib

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

datatype = MPI.FLOAT
np_dtype = dtlib.to_numpy_dtype(datatype)
itemsize = datatype.Get_size()

N = 10
NUM_RUNS = 2

if rank == 0:
    num_bytes = N * (datatype.Pack_size(count=1, comm=comm) + MPI.BSEND_OVERHEAD)
    attached_buf = bytearray(num_bytes)
    
    buf = np.empty(1, dtype=np_dtype)
    preq = comm.Bsend_init([buf, datatype], dest=1)

    for _ in range(NUM_RUNS):
        MPI.Attach_buffer(attached_buf)
        for i in range(N):
            buf[0] = i
            preq.Start()
            preq.Wait()
            print(f'Send {buf}', flush=True)
        MPI.Detach_buffer()

elif rank == 1:
    buf = np.empty(1, dtype=np_dtype)
    for _ in range(NUM_RUNS):
        for _ in range(N):
            data = comm.Recv([buf, datatype])
            print(f'Recived {buf}', flush=True)
```
What is the right value for `num_bytes`? I originally thought it would be
```python
num_bytes = datatype.Pack_size(count=N, comm=comm) + MPI.BSEND_OVERHEAD
```
but that seem too small.
Thank you very much!

Edit: Fixed bug with `Wait`

**Comment by dalcinl** [comment, marked as answer]
Using `N * (datatype.Pack_size(1, comm) + MPI.BSEND_OVERHEAD)` is correct, a per MPI-4.0 standard pp 60. You should read that material. To use the buffered-mode send, the buffer to attach should be large enough to store all the potentially in-transit messages. Anyway, I'm not 100% sure what you are really trying to do, as the rest of your code at rank 0 seems incorrect, you are issuing multiple `Start` calls on the same persistent request, without any `Wait` call in between. If you had add the missing wait calls, you may have also noticed that the `Buffer_attach` should be called BEFORE the `Bsend_init` call.

PS: This is a question about general MPI usage and semantics, and not things that are particular to Python+MPI.

**Reply by DirkKuhn** [reply]
Thank you for pointing me towards the mpi standard. I was mainly looking at "ParallelProgramming".
After some consideration, I also think that the buffering that is done in the standard send might be enough for my particular use case.
You are definitely right about the `Wait` call, I messed that up. But are you sure that the `Buffer_attach` should be called before the `Bsend_init`. I was not able to find anything on that in the standard.

PS: Yeah, you are right. Thank you for helping me anyway! Considering this, should I delete this question?

**Reply by dalcinl** [reply]
> But are you sure that the `Buffer_attach` should be called before the `Bsend_init`.

Did you try? I bet if you try hard enough (i.e. use multiple MPI implementations), you will self-answer your question.

> I was not able to find anything on that in the standard.

You are definitely right, but no standard is perfect. It is quite reasonable to assume that an MPI implementation would like to have some (attached) buffer space to store metadata about the message just posted with `Bsend_init`, even if that message is not yet active. Otherwise, what would be the point of having `MPI_BSEND_OVERHEAD` ?

> PS: Yeah, you are right. Thank you for helping me anyway! Considering this, should I delete this question?

No, no, it's OK. I was just pointing out that questions related to core MPI features and unrelated to Python/mpi4py are a bit off-topic here. The truth is, I don't really oppose people using this forum to exchange MPI tips, that would be actually welcome. The problem is that many people ask questions and very few answer. I don't like to leave questions unanswered, so I feel pressure to reply, but I don't have infinite time for everyone. In short, take my comment as a gentle request to not unnecessarily overload my *pro bono* maintenance work.
